src/Pages/ProductList/ProductList.component.js

```diff
@@ -3,15 +3,87 @@ import React, { Component } from "react";
 import "./ProductList.styles.scss";
 import ProductCard from "./ProductCard/ProductCard.component";
 
+const API = "http://localhost:3000/data/productlist/productlist.json";
+// const API = "http://10.58.4.20:8000/category?main=1"; //Backend API
+
 class ProductList extends Component {
+  state = {
+    isCategoryClick: false,
+    isSortingClick: false,
+    categories: "",
+    products: "",
+  };
+
+  handleSorting = () => {
+    const { isSortingClick } = this.state;
+    this.setState({
+      isSortingClick: !isSortingClick,
+    });
+  };
+
+  componentDidMount = async () => {
+    const response = await fetch(API);
+    // custom error handler
+    if (response.status !== 200) {
+      throw new Error("cannot fetch the data");
+    }
+    const data = await response.json();
+    this.setState({ categories: data.categories, products: data.products });
+  };
```

Comment:
```suggestion
  componentDidMount = () => {
    this.getProducts()
  };
  
  getProducts = async () => {
    try {
      const response = await fetch(API);
      if (response.status !== 200) {
        throw new Error("cannot fetch the data");
      }
      const { categories, products } = await response.json();
      this.setState({ categories, products });  
    } catch (err) {
      ...
    }
  }
```
- async/await 굿! 다만 CDM 전체를 async 함수로 선언할 필요가 없을 것 같습니다.
- 이유는 CDM의 역할과 에러 핸들링 때문입니다. 현재 작성해주신 코드는 에러를 던지고 캐치하는 부분이 없으므로 try catch 문 등을 사용해 에러를 핸들링 해야 합니다. (CDM 자체에 프로미스 체이닝을 할 수 없기 때문에)
- 아직은 데이터 호출 관련한 로직만 존재하지만 향후 그렇지 않은 로직이 섞였을 때 다른 코드도 try문 안에 작성되어야 합니다. CDM에 작성되는 코드가 복잡해지고 가독성이 떨어지게 됩니다.
- 제안드리고 싶은 방법은 데이터 호출 관련한 로직을 별도의 함수로 분리하고 해당 함수를 async로 선언, try catch 문으로 에러 핸들링을 해주는 방식이 좋을 것 같습니다.

Reply:
```js
  componentDidMount = () => {
    this.getProductList();
  };

  getProductList = async () => {
    try {
      const response = await fetch(API);
      if (response.status !== 200) {
        throw new Error("cannot fetch the data");
      }
      const { categories, products, available_sort } = await response.json();
      this.setState({ categories, products, sortings: available_sort });
    } catch (err) {
      console.log("!!error alert!!");
    }
  };
```
참고해서 다음과 같이 처리 완료했습니당! 🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻
마찬가지로 버틍을 통해 불러올때도 같은 방식으로 사용하면 될까요? 

```js
  sortingList = async (e) => {
    try {
      const { id } = e.target;
      //local test
      const response = await fetch(
        `http://localhost:3000/data/productlist/productlist_${id}.json`
      );
      //backend API
      // const response = await fetch(
      //   `http://10.58.4.20:8000/product?productList=1&sort_type=${id}`
      // );
      if (response.status !== 200) {
        throw new Error("cannot fetch the data");
      }
      const { products } = await response.json();
      this.setState({ products, activeSorting: id });
    } catch (err) {
      console.log("!!error alert!!");
    }
  };
```

Reply:
@VannsKang 네 함수 일반에 적용되는 내용입니다